imap/tls.c

```diff
@@ -1684,11 +1684,21 @@ HIDDEN int tls_start_clienttls(int readfd, int writefd,
         if (session) {
             SSL_CTX_remove_session(c_ctx, session);
         }
-        if (sess) *sess = NULL;
+        if (sess) {
+            if (*sess)
+                SSL_SESSION_free(*sess);
+            *sess = NULL;
+        }
         r = -1;
         goto done;
     }
-    if (sess) *sess = SSL_get_session(tls_conn);
+
+    /* Session may have been renegotiated; make sure we cache the right one */
+    if (sess) {
+        if (*sess)
+            SSL_SESSION_free(*sess);
+        *sess = SSL_get1_session(tls_conn);
```

Comment:
Yeah right, this one adds the refcount :) https://www.openssl.org/docs/man1.1.0/man3/SSL_get1_session.html